fundraiser/project/fundraiser-contract/src/interface.sw

```diff
@@ -125,48 +82,84 @@ abi Fundraiser {
+    /// * When the user has not pledged to the campaign represented by the `id`
+    #[storage(read, write)]
+    fn unpledge(id: u64, amount: u64);
+}
+
+abi Info {
+    /// Returns the number of unique assets that have been pledged across all campaigns
     #[storage(read)]
-    fn pledged(pledge_history_index: u64, user: Identity) -> Pledge;
+    fn asset_count() -> u64;
 
-    /// Returns the number of campaigns that the user has pledged to
+    /// Returns information about the specificed asset, specifically if it has been added and the
+    /// pledged amount
     ///
-    /// # Reverts
+    /// The user interface will not know all possible assets that the contract contains therefore
+    /// this helper method allows the interface to iterate over the asset_count to discover all assets
     ///
-    /// * When an AuthError is generated
+    /// # Arguments
+    ///
+    /// * `index` - Number from 1...asset_count
     #[storage(read)]
-    fn pledge_count(user: Identity) -> u64;
+    fn asset_info_by_count(index: u64) -> Option<AssetInfo>;
 
-    /// Returns the total number of campaigns that have been created by all users
+    /// Returns information about the specificed asset, specifically if it has been added and the
+    /// pledged amount
+    ///
+    /// # Arguments
+    ///
+    /// * `asset` - Uniquie identifier that identifies the asset
     #[storage(read)]
-    fn total_campaigns() -> u64;
+    fn asset_info_by_id(asset: ContractId) -> Option<AssetInfo>;
 
-    /// Allows a user to unpledge an amount of the campaign asset that they have pledged
-    ///
-    /// A user may have changed their mind about the amount of an asset that they have pledged
-    /// therefore they may wish to unpledge some amount of that pledge.
-    /// If they attempt to unpledge more than they have pledged then their total pledge will be returned
+    /// Returns information about the specified campaign
     ///
     /// # Arguments
     ///
     /// * `id` - Unique campaign identifier which is a number from the storage.total_campaigns range
-    /// * `amount` - The amount of asset that the user wishes to unpledge
+    #[storage(read)]
+    fn campaign_info(id: u64) -> Option<CampaignInfo>;
+
+    /// Returns information about the specified campaign for the campaign author
     ///
-    /// # Reverts
+    /// # Arguments
     ///
-    /// * When the `id` is either 0 or greater than the total number of campaigns created
-    /// * When the user attempts to unpledge after the deadline and `target_amount` have been reached
-    /// * When the user has not pledged to the campaign represented by the `id`
-    #[storage(read, write)]
-    fn unpledge(id: u64, amount: u64);
+    /// * `id` - Unique identifier which is a number starting from 1...storage.user_campaign_count
+    #[storage(read)]
+    fn campaign(campaign_history_index: u64, user: Identity) -> Option<Campaign>;
```

Comment:
No, it's convoluted. 

We need to track two things here:

1) Total number of campaigns from all users (global `CampaignInfo` data)
2) The number of unique campaigns per user
2.1) When a new campaign is created by a user the unique counter is incremented by 1

We use the unique counter and their identity to retrieve the global campaign identifier.
This allows us to query for campaigns in two ways:

1) By the actual campaign ID directly via `CampaignInfo`
2) We can check how many campaigns a user has created and use those numbers to then query the global store to get their specific campaign

If we don't do this then we need some alternative way to allow anyone to query the global data while somehow limiting the number of queries that they must make. If we only have a global store then we must iterate over the entire data. With this additional step we can ask how many campaigns a user has created and only query those limited campaigns. 

From the perspective of a UI we probably don't need this if we log and can retrieve global logs. This is a problem for contracts because they cannot get this data so they wouldn't be able to query. At least this allows contracts to also retrieve user data; but, it means they may need to loop a lot if the user has a lot of campaigns.

I don't see a good solution here except to scrap the idea completely and simplify the entire app to only contain a global store for a UI with events. 
This is probably the correct approach for us since the work here is past the MVP and has a bunch of functionality in preparation for a fully off-the-shelf application (although not yet complete).

Perhaps a future PR may just nuke most of the functionality and go with the simpler approach. cc @simonr0204

Reply:
ah gotcha thanks for the the explanation